File: lib/accuracy.ts
/*
 * Copyright (c) 2014-2021 Bjoern Kimminich.
 * SPDX-License-Identifier: MIT
 */

import logger from './logger'
import colors from 'colors/safe'
const solves: { [key: string]: { 'find it': boolean, 'fix it': boolean, attempts: { 'find it': number, 'fix it': number } } } = {}

type Phase = 'find it' | 'fix it'

export const storeFindItVerdict = (challengeKey: string, verdict: boolean) => {
  storeVerdict(challengeKey, 'find it', verdict)
}

export const storeFixItVerdict = (challengeKey: string, verdict: boolean) => {
  storeVerdict(challengeKey, 'fix it', verdict)
}

export const calculateFindItAccuracy = (challengeKey: string) => {
  return calculateAccuracy(challengeKey, 'find it')
}

export const calculateFixItAccuracy = (challengeKey: string) => {
  return calculateAccuracy(challengeKey, 'fix it')
}

export const totalFindItAccuracy = () => {
  return totalAccuracy('find it')
}

export const totalFixItAccuracy = () => {
  return totalAccuracy('fix it')
}

export const getFindItAttempts = (challengeKey: string) => {
  return solves[challengeKey] ? solves[challengeKey].attempts['find it'] : 0
}

function totalAccuracy (phase: Phase) {
  let sumAccuracy = 0
  let totalSolved = 0
  Object.entries(solves).forEach(([key, value]) => {
    if (value[phase]) {
      sumAccuracy += 1 / value.attempts[phase]
      totalSolved++
    }
  })
  return sumAccuracy / totalSolved
}

function calculateAccuracy (challengeKey: string, phase: Phase) {
  let accuracy = 0
  if (solves[challengeKey][phase]) {
    accuracy = 1 / solves[challengeKey].attempts[phase]
  }
  logger.info(`Accuracy for '${phase === 'fix it' ? 'Fix It' : 'Find It'}' phase of coding challenge ${colors.cyan(challengeKey)}: ${accuracy > 0.5 ? colors.green(accuracy.toString()) : (accuracy > 0.25 ? colors.yellow(accuracy.toString()) : colors.red(accuracy.toString()))}`)
  return accuracy
}

function storeVerdict (challengeKey: string, phase: Phase, verdict: boolean) {
  if (!solves[challengeKey]) {
    solves[challengeKey] = { 'find it': false, 'fix it': false, attempts: { 'find it': 0, 'fix it': 0 } }
  }
  if (!solves[challengeKey][phase]) {
    solves[challengeKey][phase] = verdict
    solves[challengeKey].attempts[phase]++
  }
}

File: lib/accuracy.py
import logging
import colors

logger = logging.getLogger(__name__)
solves = {}

def store_find_it_verdict(challenge_key, verdict):
    store_verdict(challenge_key, 'find it', verdict)

def store_fix_it_verdict(challenge_key, verdict):
    store_verdict(challenge_key, 'fix it', verdict)

def calculate_find_it_accuracy(challenge_key):
    return calculate_accuracy(challenge_key, 'find it')

def calculate_fix_it_accuracy(challenge_key):
    return calculate_accuracy(challenge_key, 'fix it')

def total_find_it_accuracy():
    return total_accuracy('find it')

def total_fix_it_accuracy():
    return total_accuracy('fix it')

def get_find_it_attempts(challenge_key):
    return solves[challenge_key]['attempts']['find it'] if challenge_key in solves else 0

def total_accuracy(phase):
    sum_accuracy = 0
    total_solved = 0
    for key, value in solves.items():
        if value[phase]:
            sum_accuracy += 1 / value['attempts'][phase]
            total_solved += 1
    return sum_accuracy / total_solved if total_solved > 0 else 0

def calculate_accuracy(challenge_key, phase):
    accuracy = 0
    if solves[challenge_key][phase]:
        accuracy = 1 / solves[challenge_key]['attempts'][phase]
    logger.info(f"Accuracy for '{'Fix It' if phase == 'fix it' else 'Find It'}' phase of coding challenge {colors.cyan(challenge_key)}: {colors.green(accuracy) if accuracy > 0.5 else (colors.yellow(accuracy) if accuracy > 0.25 else colors.red(accuracy))}")
    return accuracy

def store_verdict(challenge_key, phase, verdict):
    if challenge_key not in solves:
        solves[challenge_key] = {'find it': False, 'fix it': False, 'attempts': {'find it': 0, 'fix it': 0}}
    if not solves[challenge_key][phase]:
        solves[challenge_key][phase] = verdict
        solves[challenge_key]['attempts'][phase] += 1
